### PR TITLE
Bug fix for Tuple.Compare to return consistent results when comparing against a null value

### PR DESCRIPTION
**Pull request by fulghum**
`Tuple.Compare` / `Tuple.Less` returns inconsistent results when comparing different types against a null value. This caused non-null values for most types to be ordered before null values, but other types like time/date/datetime to have non-null values _after_ null values in the index. That resulted in incorrect query execution because the index range reader saw the null values and exited out before ever evaluating the non-null values. 

Fixes: https://github.com/dolthub/dolt/issues/4199

### CODE REVIEW

File: go/libraries/doltcore/sqle/enginetest/dolt_queries.go [hunk 1]
```diff
@@ -328,6 +328,24 @@ var DoltRevisionDbScripts = []queries.ScriptTest{
...
+		},
+		Assertions: []queries.ScriptTestAssertion{
+			{
+				Query:    "select count(*) from t where d is not null",
+				Expected: []sql.Row{{2}},
+			},
+			{
+				Query:    "select count(*) from t where d is null",
+				Expected: []sql.Row{{2}},
+			},
```
Comment (max-hoffman) [inline go/libraries/doltcore/sqle/enginetest/dolt_queries.go:346]: The match all expression `<=>` is often good to test when we care about nulls. Index lookups also come in two flavors, static lookups like you have above, and dynamic lookups like in indexed joins.

File: go/store/types/tuple.go [hunk 2]
```diff
@@ -690,6 +690,15 @@ func (t Tuple) TupleCompare(nbf *NomsBinFormat, otherTuple Tuple) (int, error) {
 		otherKind := otherDec.ReadKind()
 
 		if kind != otherKind {
+			// If we are comparing any type to a null type, always evaluate
```
Comment (max-hoffman) [inline go/store/types/tuple.go:693]: You might already know this, but the types package only applies to the old format. The old format also represents ranges differently and implements range scans differently. CI will fail if none of this works with the new format, so if it passes you're good.

**Review by max-hoffman — APPROVED** [review]
LGTM! just small comments

File: go/libraries/doltcore/sqle/enginetest/dolt_queries.go (hunk 1 above)
Comment (fulghum) [reply go/libraries/doltcore/sqle/enginetest/dolt_queries.go:346]: Thanks for the tips! I added assertions for the null-safe equals operator and for an indexed join query.

**Pull request merged**